common/etc/nginx/include/s3gateway.js

```diff
@@ -643,10 +683,108 @@ function _require_env_var(envVarName) {
     }
 }
 
+var maxValidityOffset = 270000;
+
+async function fetchCredentials(r) {
+    var current = readCredentials();
+    if (current) {
+        var exp = new Date(current.expiration).getTime() - maxValidityOffset;
+        if (now.getTime() < exp) {
+            r.return(200);
+            return;
+        }
+    }
+    _debug_log(r, 'Cached credentials are expired or not present, requesting new ones');
+
+    var credentials;
+    if (process.env['S3_ACCESS_KEY_ID']) {
+        credentials = {
+            accessKeyId: process.env['S3_ACCESS_KEY_ID'],
+            secretAccessKey: process.env['S3_SECRET_KEY'],
+            sessionToken: null,
+        };
+    } else if (process.env['AWS_CONTAINER_CREDENTIALS_RELATIVE_URI']) {
+        var uri = ecsCredentialsBaseUri + process.env['AWS_CONTAINER_CREDENTIALS_RELATIVE_URI']
+        try {
+            credentials = await _fetchEcsRoleCredentials(uri);
+        } catch (e) {
+            _debug_log(r, 'Could not load ECS task role credentials: ' + JSON.stringify(e));
+            r.return(500);
+            return;
+        }
+    } else {
+        try {
+            credentials = await _fetchEC2RoleCredentials();
+        } catch (e) {
+            _debug_log(r, 'Could not load EC2 task role credentials: ' + JSON.stringify(e));
+            r.return(500);
+            return;
+        }
+    }
+    try {
+        fs.writeFileSync('/tmp/credentials.json', JSON.stringify(credentials));
+    } catch (e) {
+        _debug_log(r, 'Could not write credentials file: ' + JSON.stringify(e));
+        r.return(500);
+        return;
+    }
+    r.return(200);
+}
+
+async function _fetchEcsRoleCredentials(credentialsUri) {
+    var resp = await ngx.fetch(credentialsUri);
+    if (!resp.ok) {
+        throw 'Credentials endpoint response was not ok.';
+    }
+    var creds = await resp.json();
+
+    return {
+        accessKeyId: creds.AccessKeyId,
+        secretAccessKey: creds.SecretAccessKey,
+        sessionToken: creds.Token,
+        expiration: creds.Expiration,
+    };
+}
+
+async function _fetchEC2RoleCredentials() {
+    var tokenResp = await ngx.fetch(ec2ImdsTokenEndpoint, {
+        headers: {
+            'x-aws-ec2-metadata-token-ttl-seconds': '21600',
+        },
+        method: 'PUT',
+    });
+    var token = await tokenResp.text();
+    var resp = await ngx.fetch(ec2ImdsSecurityCredentialsEndpoint, {
+        headers: {
+            'x-aws-ec2-metadata-token': token,
+        },
+    });
+    // TODO: I'm lazy, but this can be a list of roles, however, for sake of simplicity, let's assume it's just one
```

Comment:
Let's change this TODO to an explanation that we are only doing a single role right now and then let's create a ticket for it.

Reply:
I looked it up when doing this code, the thing is:
- the AWS SDK in Go, iirc, reads each line as a role and selects the first one by default.
- an EC2 instance _can_ only have one role attached

Therefore, I think it might be just a safeguard in the SDK, as maybe the IMDS can return multiple roles. However, in the current scenario (EC2 instances) there can only be one role at any time. I change the comment to reflect that, though 👍

Reply:
Meaning also: I don't think we need a separate issue for that, tbh. There's currently no use-case, I know, which uses IMDS with multiple roles.